### ISSUE TITLE
ThinLTO lit tests fail with bfd linker

### ISSUE THREAD

**Issue by Fatmice** [ldc-developers/ldc#2278]
I built and install llvm 4.0.1 with
`cmake -DCMAKE_INSTALL_PREFIX=../llvm-test -DLLVM_TARGETS_TO_BUILD="X86;AArch64;ARM;PowerPC" -DLLVM_BINUTILS_INCDIR=/usr/include -DCMAKE_BUILD_TYPE=Release -DLLVM_LIBDIR_SUFFIX=64 ../llvm`
Then built and test ldc origin/master with
`cmake -DCMAKE_INSTALL_PREFIX=../ldc-test -DINCLUDE_PATH=import -DCMAKE_BUILD_TYPE=Release -DLIB_SUFFIX=64 -DLDC_INSTALL_LTOPLUGIN=ON -DLDC_INSTALL_LLVM_RUNTIME_LIBS=ON -DLLVM_ROOT_DIR=../llvm-test -DD_COMPILER=../install/ldc-0.17.4-bootstrap/bin/ldmd2 ../ldc`

All tests passed except for two of the lit tests.
linking/thinlto_modulecdtors.d (134 of 153)
linking/thinlto_modulecdtors_2.d (144 of 153)
Here's [complete log](https://gist.githubusercontent.com/Fatmice/ef8dd6750b1a9cb98d03f193f5d96bb1/raw/4d5d1dad97bdc9f7e9a5675d04fa57c8f05cb28f/LastTest.log)

The relevant portions
```
FAIL: LDC :: linking/thinlto_modulecdtors.d (134 of 153)
******************** TEST 'LDC :: linking/thinlto_modulecdtors.d' FAILED ********************
Script:
--
/home/quyen/D/ldc-build/bin/ldc2 -flto=thin -O3 -run /home/quyen/D/ldc/tests/linking/thinlto_modulecdtors.d | FileCheck /home/quyen/D/ldc/tests/linking/thinlto_modulecdtors.d
--
Exit Code: 2

Command Output (stdout):
--
$ "/home/quyen/D/ldc-build/bin/ldc2" "-flto=thin" "-O3" "-run" "/home/quyen/D/ldc/tests/linking/thinlto_modulecdtors.d"
# command stderr:
Error: /tmp/thinlto_modulecdtors-93a9e0a-1f491d failed with status: -2
Error: message: Illegal instruction (core dumped)
Error: program received signal 2 (Interrupt)

error: command failed with exit status: 2
$ "FileCheck" "/home/quyen/D/ldc/tests/linking/thinlto_modulecdtors.d"
# command stderr:
FileCheck error: '-' is empty.
FileCheck command line:  FileCheck /home/quyen/D/ldc/tests/linking/thinlto_modulecdtors.d

error: command failed with exit status: 2

--

********************
UNSUPPORTED: LDC :: sanitizers/asan_dynalloc.d (135 of 153)
UNSUPPORTED: LDC :: sanitizers/asan_interface.d (136 of 153)
UNSUPPORTED: LDC :: sanitizers/asan_stackoverflow.d (137 of 153)
PASS: LDC :: sanitizers/fsanitize_address.d (138 of 153)
PASS: LDC :: linking/thinlto_pgo.d (139 of 153)
PASS: LDC :: sanitizers/fsanitize_blacklist.d (140 of 153)
PASS: LDC :: sanitizers/fsanitize_error_diagnostic.d (141 of 153)
PASS: LDC :: sanitizers/fsanitize_coverage.d (142 of 153)
PASS: LDC :: sanitizers/fsanitize_thread.d (143 of 153)
FAIL: LDC :: linking/thinlto_modulecdtors_2.d (144 of 153)
******************** TEST 'LDC :: linking/thinlto_modulecdtors_2.d' FAILED ********************
Script:
--
/home/quyen/D/ldc-build/bin/ldc2 -flto=thin -O3 /home/quyen/D/ldc/tests/linking/inputs/thinlto_ctor.d -run /home/quyen/D/ldc/tests/linking/thinlto_modulecdtors_2.d | FileCheck --check-prefix=EXECUTE /home/quyen/D/ldc/tests/linking/thinlto_modulecdtors_2.d
--
Exit Code: 2

Command Output (stdout):
--
$ "/home/quyen/D/ldc-build/bin/ldc2" "-flto=thin" "-O3" "/home/quyen/D/ldc/tests/linking/inputs/thinlto_ctor.d" "-run" "/home/quyen/D/ldc/tests/linking/thinlto_modulecdtors_2.d"
# command stderr:
Error: /tmp/thinlto_modulecdtors_2-939d98e-2558f3 failed with status: -2
Error: message: Illegal instruction (core dumped)
Error: program received signal 2 (Interrupt)

error: command failed with exit status: 2
$ "FileCheck" "--check-prefix=EXECUTE" "/home/quyen/D/ldc/tests/linking/thinlto_modulecdtors_2.d"
# command stderr:
FileCheck error: '-' is empty.
FileCheck command line:  FileCheck --check-prefix=EXECUTE /home/quyen/D/ldc/tests/linking/thinlto_modulecdtors_2.d

error: command failed with exit status: 2

--

********************
UNSUPPORTED: LDC :: sanitizers/fuzz_basic.d (145 of 153)
UNSUPPORTED: LDC :: sanitizers/fuzz_asan.d (146 of 153)
UNSUPPORTED: LDC :: sanitizers/link_fuzzer.d (147 of 153)
PASS: LDC :: sanitizers/fsanitize_fuzzer.d (148 of 153)
PASS: LDC :: semantic/target_traits_diag.d (149 of 153)
PASS: LDC :: semantic/dcompute.d (150 of 153)
PASS: LDC :: semantic/target_traits.d (151 of 153)
PASS: LDC :: linking/ir2obj_cache_pruning2.d (152 of 153)
PASS: LDC :: tools/ldc_prune_cache_1.d (153 of 153)
Testing Time: 20.31s
********************
Failing Tests (2):
    LDC :: linking/thinlto_modulecdtors.d
    LDC :: linking/thinlto_modulecdtors_2.d

  Expected Passes    : 128
  Unsupported Tests  : 23
  Unexpected Failures: 2
<end of output>
Test time =  20.59 sec
----------------------------------------------------------
Test Failed.
"lit-tests" end time: Aug 16 16:24 CDT
"lit-tests" time elapsed: 00:00:20
----------------------------------------------------------

End testing: Aug 16 16:24 CDT
```

I then make install the ldc build and try to repeat those two tests
[linking/thinlto_modulecdtors.d](https://gist.github.com/Fatmice/37892dd004e4dc812017f9e028bea752#file-linking-thinlto_modulecdtors-d)
[linking/thinlto_modulecdtors_2.d](https://gist.github.com/Fatmice/37892dd004e4dc812017f9e028bea752#file-linking-thinlto_modulecdtors_2-d)
Running the executables from these two always resulted in Illegal instruction (core dumped) when -flto=thin is used but -flto=full will run without problem.
[linking/thinlto_modulecdtors.d](https://gist.github.com/Fatmice/1f243c049de1f1fd2efd1338530a6c7e#file-linking-thinlto_modulecdtors-d)
[linking/thinlto_modulecdtors_2.d](https://gist.github.com/Fatmice/1f243c049de1f1fd2efd1338530a6c7e#file-linking-thinlto_modulecdtors_2-d)

What is the issue here? @JohanEngelen ?

**Comment by Fatmice** [comment]
What else should I do to help debug this?

**Comment by JohanEngelen** [comment]
This looks like a ThinLTO bug (in LLVM). Ideas to help debug:
- system details, OS, processor, ...
- reduce to a tiny testcase: I think you are not able to create _any_ executable with -flto=thin, right? (it's not due to something special in the Lit tests)
- Can you test whether things work with clang?
- Try to figure out if it is really an illegal instruction, or is it a jump/call to invalid address? I think it's very unlikely, given that we pass the CPU in LLVM IR (IIRC), but it could be that the LTO plugin defaults to a newer CPU than your system?

**Comment by JohanEngelen** [comment]
This may be good to try too:
- compile the super tiny program (try C-like `main` directly, without having to link to druntime/phobos) to LLVM IR  (`-output-ll`)
- Use llvm llc to get a thinlto object file
- Link with `lld`.
- see if the executable runs.
If you still get errors with `-flto=thin` but not without it, then we know for sure it is a bug somewhere in LLVM.

**Comment by JohanEngelen** [comment]
For some extra ideas: https://github.com/ldc-developers/ldc/issues/2280

Thanks for taking the time to find out what the problem is. (I know it takes a lot of time :( )

**Comment by Fatmice** [comment]
> system details, OS, processor, ...
reduce to a tiny testcase: I think you are not able to create any executable with -flto=thin, right? (it's not due to something special in the Lit tests)

```[quyen@thinkpad ldc-test]$ ./bin/ldc2 --version
LDC - the LLVM D compiler (1.4.0git-1a3391e):
  based on DMD v2.074.1 and LLVM 4.0.1
  built with LDC - the LLVM D compiler (0.17.4)
  Default target: x86_64-unknown-linux-gnu
  Host CPU: haswell
  http://dlang.org - http://wiki.dlang.org/LDC

  Registered Targets:
    aarch64    - AArch64 (little endian)
    aarch64_be - AArch64 (big endian)
    arm        - ARM
    arm64      - ARM64 (little endian)
    armeb      - ARM (big endian)
    ppc32      - PowerPC 32
    ppc64      - PowerPC 64
    ppc64le    - PowerPC 64 LE
    thumb      - Thumb
    thumbeb    - Thumb (big endian)
    x86        - 32-bit X86: Pentium-Pro and above
    x86-64     - 64-bit X86: EM64T and AMD64
[quyen@thinkpad ldc-test]$ cat /proc/cpuinfo 
processor	: 0
vendor_id	: GenuineIntel
cpu family	: 6
model		: 69
model name	: Intel(R) Core(TM) i3-4005U CPU @ 1.70GHz
stepping	: 1
microcode	: 0x20
cpu MHz		: 1000.036
cache size	: 3072 KB
physical id	: 0
siblings	: 4
core id		: 0
cpu cores	: 2
apicid		: 0
initial apicid	: 0
fpu		: yes
fpu_exception	: yes
cpuid level	: 13
wp		: yes
flags		: fpu vme de pse tsc msr pae mce cx8 apic sep mtrr pge mca cmov pat pse36 clflush dts acpi mmx fxsr sse sse2 ss ht tm pbe syscall nx pdpe1gb rdtscp lm constant_tsc arch_perfmon pebs bts rep_good nopl xtopology nonstop_tsc cpuid aperfmperf pni pclmulqdq dtes64 monitor ds_cpl vmx est tm2 ssse3 sdbg fma cx16 xtpr pdcm pcid sse4_1 sse4_2 movbe popcnt tsc_deadline_timer aes xsave avx f16c rdrand lahf_lm abm cpuid_fault epb tpr_shadow vnmi flexpriority ept vpid fsgsbase tsc_adjust bmi1 avx2 smep bmi2 erms invpcid xsaveopt dtherm ida arat pln pts
bugs		:
bogomips	: 3392.30
clflush size	: 64
cache_alignment	: 64
address sizes	: 39 bits physical, 48 bits virtual
power management:
[quyen@thinkpad ldc-test]$ uname -a
Linux thinkpad 4.12.5-300.fc26.x86_64 #1 SMP Mon Aug 7 15:27:25 UTC 2017 x86_64 x86_64 x86_64 GNU/Linux
```
Also I've found that -flto=thin does work but I can not pass -O# of any kind.  It will cause either segmentation fault or illegal instruction.  -flto=full will work with optimizations of any kind.

I'm currently using this small test case
```
[quyen@thinkpad lit-test]$ cat app.d 
import std.stdio;

void main() {
    int a = int.max;
    int b;
    writeln("Max Int: ", a);
    writeln("Overflow: ", a+a);
    writeln("Overflow: ", a+a+a);
    writeln("Overflow: ", a+a+a+a);
    writeln("Overflow: ", a+a+a+a+a);
}
```

> 
    compile the super tiny program (try C-like main directly, without having to link to druntime/phobos) to LLVM IR (-output-ll)
    Use llvm llc to get a thinlto object file
    Link with lld.
    see if the executable runs.
    If you still get errors with -flto=thin but not without it, then we know for sure it is a bug somewhere in LLVM.
Looks complicated...  So compile with llvm-gcc?

**Comment by JohanEngelen** [comment]
Minimizing the testcase to be as small as possible will help analyzing the IR (if need be). Does `int main() { return 0; }` also break?

**Comment by Fatmice** [comment]
Yep, with any kind of optimization.

**Comment by Fatmice** [comment]
../ldc-test/bin/ldc2 -flto=thin -output-ll simple.d
[simple.ll](https://gist.github.com/Fatmice/4cf61a435daed53663b081a78c6ba640)
../llvm-test/bin/llc -O3 -filetype=obj simple.ll
[simple.o.zip](https://github.com/ldc-developers/ldc/files/1243177/simple.o.zip)
Now I'm stuck on the ../lvm-test/bin/lc.lld as that issued way too many symbols errors.

**Comment by Fatmice** [comment]
[Diff between simple.ll and -O3 simple.ll](https://gist.github.com/Fatmice/4cf61a435daed53663b081a78c6ba640#file-diff)

**Comment by Fatmice** [comment]
@JohanEngelen any insight?  Maybe LDC 1.4 and a newer LLVM will fix this issue?

**Comment by JohanEngelen** [comment]
Sorry, haven't looked at this any more.
(the linker errors are because you need to link with the C runtime [you can figure out what LDC links with by adding `-v` to LDC cmdline when linking])

I think when you use `-link-debuglib`, you can see more debug info (line number hopefully) where the crash happens. I think you are hitting and assert(0) in druntime code.

**Comment by Fatmice** [comment]
@JohanEngelen 
The cmd line had -v in it in the links of in the bottom of the first post.
The binary **only crashes** if -flto=thin is used **with** -Ox (with optimization of any kind).  **Why is that?**
If it didn't link with the c runtime libraries then it wouldn't even run with or without optimization...

**Comment by Fatmice** [comment]
```
[quyen@thinkpad lit-test]$ ../ldc-test/bin/ldc2 -v -flto=thin -O3 -link-debuglib -run simple.d 
binary    /home/quyen/D/ldc-test/bin/ldc2
version   1.4.0git-beb1cd5 (DMD v2.074.1, LLVM 4.0.1)
config    /home/quyen/D/ldc-test/etc/ldc2.conf (x86_64-unknown-linux-gnu)
predefs   LDC all D_Version2 assert X86_64 D_InlineAsm_X86_64 D_HardFloat LittleEndian D_LP64 D_PIC linux Posix CRuntime_Glibc LDC_LLVM_400
parse     simple
importall simple
import    object	(/home/quyen/D/ldc-test/include/ldc/object.d)
semantic  simple
entry     main      	simple.d
semantic2 simple
semantic3 simple
code      simple
/usr/bin/gcc simple-0bca230.o -o /tmp/simple-0bca230-e7783f -Wl,-plugin,/home/quyen/D/ldc-test/lib64/LLVMgold-ldc.so -Wl,-plugin-opt=thinlto -Wl,-plugin-opt=O3 -Wl,-plugin-opt=-function-sections -Wl,-plugin-opt=-data-sections -L/home/quyen/D/ldc-test/bin/../lib64 -lphobos2-ldc-debug -ldruntime-ldc-debug -Wl,--gc-sections -lrt -ldl -lpthread -lm -m64 
/tmp/simple-0bca230-e7783f 
Error: /tmp/simple-0bca230-e7783f failed with status: -2
Error: message: Segmentation fault (core dumped)
Error: program received signal 2 (Interrupt)
```

**Comment by kinke** [comment]
As Johan mentioned and is now visible in your backtrace, you seem to be hitting this assert [here](https://github.com/ldc-developers/druntime/blob/ldc/src/rt/sections_elf_shared.d#L498). We've just had a similar report, https://github.com/ldc-developers/ldc/issues/2280#issuecomment-323552845, which isn't LTO-specific though. It appears specific to shared runtime libs though.

**Comment by Fatmice** [comment]
So what to do? I do not have issues when compiling with flto=full and optimization.  Without the flto flag, it also compiles just fine.

**Comment by kinke** [comment]
You could check if your `-flto=thin` binary has other runtime dependencies on shared objects than the `-flto=full` one. The failing assert checks that there's only a single D binary for the whole process.
As to our CI setup, you'll have to wade through our CI scripts, e.g., the [Travis one](https://github.com/ldc-developers/ldc/blob/master/.travis.yml), or check their logs, due to manifold details.

**Comment by Fatmice** [comment]
> You could check if your -flto=thin binary has other runtime dependencies on shared objects than the -flto=full one.

Using ldd? It shown above that there isn't any run time dependencies differences...

> As to our CI setup, you'll have to wade through our CI scripts, e.g., the Travis one, or check their logs, due to manifold details.

I read those multiple times already, nothing outstanding at all.  I compiled LLVM with gcc and not with clang.

**Comment by Fatmice** [comment]
```
-rwxrwxr-x. 1 quyen 1343960 Sep  4 19:21 simple-fullWOP*
-rwxrwxr-x. 1 quyen 1344032 Sep  4 19:20 simple-thinWOP*
-rwxrwxr-x. 1 quyen 1343936 Sep  4 19:23 simple-thinNOP*
```
superficially, the binaries are different in size

**Comment by kinke** [comment]
> Using ldd? It shown above that there isn't any run time dependencies differences...

Between optimized and non-optimized builds, yes, I forgot that it's working if not optimized.

> I read those multiple times already, nothing outstanding at all. I compiled LLVM with gcc and not with clang.

Our CI bots are based on Ubuntu 14.04. Try to build it in a Docker container may be worth a try.

**Comment by Fatmice** [comment]
> Between optimized and non-optimized builds, yes, I forgot that it's working if not optimized.

It's working with -flto=full -O3 (or any of the Os)
It's working with -flto=thin without any Os
It's working without flto and any Os.

**Comment by kinke** [comment]
Another thing to try would be setting a breakpoint in _d_dso_registry and see if it' somehow hit multiple times, and if so, from where.

**Comment by Fatmice** [comment]
```
[quyen@thinkpad lit-test]$ ls
total 5284
-rw-rw-r--. 1 quyen     236 Aug 16 18:18 app.d
-rw-rw-r--. 1 quyen      46 Aug 22 14:01 simple.c
-rw-rw-r--. 1 quyen      25 Aug 22 13:56 simple.d
-rwxrwxr-x. 1 quyen 1343960 Sep  4 19:35 simple-fullNOP*
-rw-rw-r--. 1 quyen    2512 Sep  4 19:35 simple-fullNOP.o
-rwxrwxr-x. 1 quyen 1343960 Sep  4 19:35 simple-fullWOP*
-rw-rw-r--. 1 quyen    2768 Sep  4 19:35 simple-fullWOP.o
-rwxrwxr-x. 1 quyen 1343936 Sep  4 19:35 simple-thinNOP*
-rw-rw-r--. 1 quyen    2708 Sep  4 19:35 simple-thinNOP.o
-rwxrwxr-x. 1 quyen 1344032 Sep  4 19:35 simple-thinWOP*
-rw-rw-r--. 1 quyen    2956 Sep  4 19:35 simple-thinWOP.o
[quyen@thinkpad lit-test]$ ldd simple-fullNOP
	linux-vdso.so.1 (0x00007ffc3b395000)
	librt.so.1 => /lib64/librt.so.1 (0x00007fc44ff11000)
	libdl.so.2 => /lib64/libdl.so.2 (0x00007fc44fd0d000)
	libpthread.so.0 => /lib64/libpthread.so.0 (0x00007fc44faee000)
	libm.so.6 => /lib64/libm.so.6 (0x00007fc44f7d8000)
	libgcc_s.so.1 => /lib64/libgcc_s.so.1 (0x00007fc44f5c1000)
	libc.so.6 => /lib64/libc.so.6 (0x00007fc44f1f0000)
	/lib64/ld-linux-x86-64.so.2 (0x00007fc450119000)
[quyen@thinkpad lit-test]$ ldd simple-fullWOP
	linux-vdso.so.1 (0x00007ffe6ec4d000)
	librt.so.1 => /lib64/librt.so.1 (0x00007f7b96a0a000)
	libdl.so.2 => /lib64/libdl.so.2 (0x00007f7b96806000)
	libpthread.so.0 => /lib64/libpthread.so.0 (0x00007f7b965e7000)
	libm.so.6 => /lib64/libm.so.6 (0x00007f7b962d1000)
	libgcc_s.so.1 => /lib64/libgcc_s.so.1 (0x00007f7b960ba000)
	libc.so.6 => /lib64/libc.so.6 (0x00007f7b95ce9000)
	/lib64/ld-linux-x86-64.so.2 (0x00007f7b96c12000)
[quyen@thinkpad lit-test]$ ldd simple-thinNOP
	linux-vdso.so.1 (0x00007ffedbd98000)
	librt.so.1 => /lib64/librt.so.1 (0x00007f1507cd7000)
	libdl.so.2 => /lib64/libdl.so.2 (0x00007f1507ad3000)
	libpthread.so.0 => /lib64/libpthread.so.0 (0x00007f15078b4000)
	libm.so.6 => /lib64/libm.so.6 (0x00007f150759e000)
	libgcc_s.so.1 => /lib64/libgcc_s.so.1 (0x00007f1507387000)
	libc.so.6 => /lib64/libc.so.6 (0x00007f1506fb6000)
	/lib64/ld-linux-x86-64.so.2 (0x00007f1507edf000)
[quyen@thinkpad lit-test]$ ldd simple-thinWOP
	linux-vdso.so.1 (0x00007ffcc2792000)
	librt.so.1 => /lib64/librt.so.1 (0x00007f079a42b000)
	libdl.so.2 => /lib64/libdl.so.2 (0x00007f079a227000)
	libpthread.so.0 => /lib64/libpthread.so.0 (0x00007f079a008000)
	libm.so.6 => /lib64/libm.so.6 (0x00007f0799cf2000)
	libgcc_s.so.1 => /lib64/libgcc_s.so.1 (0x00007f0799adb000)
	libc.so.6 => /lib64/libc.so.6 (0x00007f079970a000)
	/lib64/ld-linux-x86-64.so.2 (0x00007f079a633000)
[quyen@thinkpad lit-test]$ 
```
I do not see any differences.  There are clearly binary size differences.

> Another thing to try would be setting a breakpoint in _d_dso_registry and see if it' somehow hit multiple times, and if so, from where.
I do not know how to do that.

**Comment by thewilsonator** [comment]
```
$lldb -- program -args
(lldb) b _d_dso_registry
(lldb) run
```
and then see how many times the breakpoint is hit.

**Comment by kinke** [comment]
> I do not know how to do that.

Google for `gdb cheat sheet`. ;) - OTOH:
```
gdb --args ./myExecutable
b sections_elf_shared.d:498
r
```

**Comment by Fatmice** [comment]
```
Starting program: /home/quyen/D/lit-test/simple-thinWOP 
[Thread debugging using libthread_db enabled]
Using host libthread_db library "/lib64/libthread_db.so.1".

Breakpoint 1, _d_dso_registry (arg=0x0) at sections_elf_shared.d:425
425	extern(C) void _d_dso_registry(void* arg)
(gdb) c
Continuing.

Breakpoint 1, _d_dso_registry (
    arg=0x7f1e90fc1d73 <_dl_init+259> "I\203\306\bL;4$u\353\203\353\001\203\373\377\017\205C\377\377\377H\215\005\360s!")
    at sections_elf_shared.d:425
425	extern(C) void _d_dso_registry(void* arg)
(gdb) c
Continuing.

Program received signal SIGSEGV, Segmentation fault.
0x0000000000417b9c in gc_malloc (sz=80, ba=0, ti=0x64fad0 <ClassInfo for core.exception.AssertError>) at proxy.d:117
117	        return instance.malloc(sz, ba, ti);
(gdb) c
Continuing.

Program terminated with signal SIGSEGV, Segmentation fault.
The program no longer exists.
(gdb) 
```
It entered it twice.  This one is compiled with -flto=thin -O3

**Comment by Fatmice** [comment]
```
Reading symbols from simple-thinWOP...done.
(gdb) b sections_elf_shared.d:498
Breakpoint 1 at 0x40b22b: file sections_elf_shared.d, line 498.
(gdb) r
Starting program: /home/quyen/D/lit-test/simple-thinWOP 
[Thread debugging using libthread_db enabled]
Using host libthread_db library "/lib64/libthread_db.so.1".

Breakpoint 1, _d_dso_registry (arg=0x7fffffffc040 "\001") at sections_elf_shared.d:498
498	                _loadedDSOs.empty ||
(gdb) c
Continuing.

Breakpoint 1, _d_dso_registry (arg=0x7fffffffc020 "\001") at sections_elf_shared.d:498
498	                _loadedDSOs.empty ||
(gdb) c
Continuing.

Program received signal SIGSEGV, Segmentation fault.
0x0000000000417b9c in gc_malloc (sz=80, ba=0, ti=0x64fad0 <ClassInfo for core.exception.AssertError>) at proxy.d:117
117	        return instance.malloc(sz, ba, ti);
(gdb) c
Continuing.

Program terminated with signal SIGSEGV, Segmentation fault.
The program no longer exists.
(gdb) 
```

**Comment by Fatmice** [comment]
```
Reading symbols from simple-thinNOP...done.
(gdb) b _d_dso_registry
Breakpoint 1 at 0x40af20: file sections_elf_shared.d, line 425.
(gdb) b sections_elf_shared.d:498
Breakpoint 2 at 0x40b1fb: file sections_elf_shared.d, line 498.
(gdb) r
Starting program: /home/quyen/D/lit-test/simple-thinNOP 
[Thread debugging using libthread_db enabled]
Using host libthread_db library "/lib64/libthread_db.so.1".

Breakpoint 1, _d_dso_registry (
    arg=0x7f4c1bfe2d73 <_dl_init+259> "I\203\306\bL;4$u\353\203\353\001\203\373\377\017\205C\377\377\377H\215\005\360s!")
    at sections_elf_shared.d:425
425	extern(C) void _d_dso_registry(void* arg)
(gdb) c
Continuing.

Breakpoint 2, _d_dso_registry (arg=0x7fffffffc020 "\001") at sections_elf_shared.d:498
498	                _loadedDSOs.empty ||
(gdb) c
Continuing.

Breakpoint 1, _d_dso_registry (arg=0x0) at sections_elf_shared.d:425
425	extern(C) void _d_dso_registry(void* arg)
(gdb) c
Continuing.
[Inferior 1 (process 15782) exited normally]
```
This one is compiled with only -flto=thin

**Comment by Fatmice** [comment]
```
(gdb) file simple-fullWOP
Load new symbol table from "simple-fullWOP"? (y or n) y
Reading symbols from simple-fullWOP...done.
(gdb) r
Starting program: /home/quyen/D/lit-test/simple-fullWOP 
[Thread debugging using libthread_db enabled]
Using host libthread_db library "/lib64/libthread_db.so.1".

Breakpoint 1, _d_dso_registry (arg=0x7f9be0831950 "\360\330a\340\233\177") at sections_elf_shared.d:425
425	extern(C) void _d_dso_registry(void* arg)
(gdb) c
Continuing.

Breakpoint 2, _d_dso_registry (arg=0x7fffffffc030 "\001") at sections_elf_shared.d:498
498	                _loadedDSOs.empty ||
(gdb) c
Continuing.

Breakpoint 1, _d_dso_registry (arg=0x0) at sections_elf_shared.d:425
425	extern(C) void _d_dso_registry(void* arg)
(gdb) c
Continuing.
[Inferior 1 (process 15805) exited normally]
```
This one is compiled with -flto=full -O3
(gdb) file simple-fullNOP
Load new symbol table from "simple-fullNOP"? (y or n) y
Reading symbols from simple-fullNOP...done.
(gdb) r
Starting program: /home/quyen/D/lit-test/simple-fullNOP 
[Thread debugging using libthread_db enabled]
Using host libthread_db library "/lib64/libthread_db.so.1".

Breakpoint 1, _d_dso_registry (
    arg=0x7f92d4b2dd73 <_dl_init+259> "I\203\306\bL;4$u\353\203\353\001\203\373\377\017\205C\377\377\377H\215\005\360s!")
    at sections_elf_shared.d:425
425	extern(C) void _d_dso_registry(void* arg)
(gdb) c
Continuing.

Breakpoint 2, _d_dso_registry (arg=0x7fffffffc020 "\001") at sections_elf_shared.d:498
498	                _loadedDSOs.empty ||
(gdb) c
Continuing.

Breakpoint 1, _d_dso_registry (arg=0x0) at sections_elf_shared.d:425
425	extern(C) void _d_dso_registry(void* arg)
(gdb) c
Continuing.
[Inferior 1 (process 15808) exited normally]
```
This one is compiled with only -flto=full

**Comment by kinke** [comment]
So apparently using thin LTO somehow makes your executable try to register itself at least twice. Note that hitting `_d_dso_registry()` twice is normal (register at startup + unregister on termination). I can't tell you more at the moment.

**Comment by Fatmice** [comment]
> 

So apparently using thin LTO somehow makes your executable try to register itself at least twice. Note that hitting _d_dso_registry() twice is normal (register at startup + unregister on termination). I can't tell you more at the moment.
--

Only if optimization is used with flto=thin though.  Otherwise it terminates without issues.

**Comment by Fatmice** [comment]
So out of curiosity I compiled with -O3 and without -flto.  It is the LTO-thin that is the issue as this program terminates without issues.  Something is going on at _loadedDSOs.empty when -O3 and -flto=thin is used.
```
[quyen@thinkpad lit-test]$ ../ldc-test/bin/ldc2 -v -O3 simple.d
binary    /home/quyen/D/ldc-test/bin/ldc2
version   1.4.0git-beb1cd5 (DMD v2.074.1, LLVM 4.0.1)
config    /home/quyen/D/ldc-test/etc/ldc2.conf (x86_64-unknown-linux-gnu)
predefs   LDC all D_Version2 assert X86_64 D_InlineAsm_X86_64 D_HardFloat LittleEndian D_LP64 D_PIC linux Posix CRuntime_Glibc LDC_LLVM_400
parse     simple
importall simple
import    object	(/home/quyen/D/ldc-test/include/ldc/object.d)
semantic  simple
entry     main      	simple.d
semantic2 simple
semantic3 simple
code      simple
/usr/bin/gcc simple.o -o simple -L/home/quyen/D/ldc-test/bin/../lib64 -lphobos2-ldc -ldruntime-ldc -Wl,--gc-sections -lrt -ldl -lpthread -lm -m64 
[quyen@thinkpad lit-test]$ ls
total 360
-rw-rw-r--. 1 quyen    236 Aug 16 18:18 app.d
-rwxrwxr-x. 1 quyen 350864 Sep  5 17:03 simple*
-rw-rw-r--. 1 quyen     46 Aug 22 14:01 simple.c
-rw-rw-r--. 1 quyen     25 Aug 22 13:56 simple.d
-rw-rw-r--. 1 quyen   4200 Sep  5 17:03 simple.o
[quyen@thinkpad lit-test]$ gdb simple 
GNU gdb (GDB) Fedora 8.0-24.fc26
Copyright (C) 2017 Free Software Foundation, Inc.
License GPLv3+: GNU GPL version 3 or later <http://gnu.org/licenses/gpl.html>
This is free software: you are free to change and redistribute it.
There is NO WARRANTY, to the extent permitted by law.  Type "show copying"
and "show warranty" for details.
This GDB was configured as "x86_64-redhat-linux-gnu".
Type "show configuration" for configuration details.
For bug reporting instructions, please see:
<http://www.gnu.org/software/gdb/bugs/>.
Find the GDB manual and other documentation resources online at:
<http://www.gnu.org/software/gdb/documentation/>.
For help, type "help".
Type "apropos word" to search for commands related to "word"...
Reading symbols from simple...(no debugging symbols found)...done.
(gdb) b _d_dso_registry
Breakpoint 1 at 0x407540
(gdb) b sections_elf_shared.d:498
No symbol table is loaded.  Use the "file" command.
Make breakpoint pending on future shared library load? (y or [n]) y
Breakpoint 2 (sections_elf_shared.d:498) pending.
(gdb) r
Starting program: /home/quyen/D/lit-test/simple 
[Thread debugging using libthread_db enabled]
Using host libthread_db library "/lib64/libthread_db.so.1".

Breakpoint 1, 0x0000000000407540 in _d_dso_registry ()
(gdb) c
Continuing.

Breakpoint 1, 0x0000000000407540 in _d_dso_registry ()
(gdb) c
Continuing.
[Inferior 1 (process 20113) exited normally]
(gdb) 
```

**Comment by kinke** [comment]
I'm hitting this with LLVM 5.0 on Ubuntu 16.04 as well.

**Comment by Fatmice** [comment]
Welcome to the party ;)

Not to blame anyone but this is surely a LDC problem.  I'm building LLVM trunk at the moment (been doing so for the last 3 days) to build PGO bootstrap of clang and gang with all the bells and whistles (LTO=thin, LLD) and haven't yet have any issues with thin link time optimization.  Thus the problem must be purely on LDC.  Once the bootstrap is done, I will use that to build the newly released LLVM 5.0.0 and LLVM trunk for use with LDC.  If this problem still occur, then clearly it is not LLVM

**Comment by JohanEngelen** [comment]
What may help is passing `-L-save-temps`, such that the linker outputs a few intermediate files (bitcode) and you can study the bitcode after ThinLTO inlining across modules and other optimizations.

**Comment by Fatmice** [comment]
> What may help is passing -L-save-temps, such that the linker outputs a few intermediate files (bitcode) and you can study the bitcode after ThinLTO inlining across modules and other optimizations.

I did give the intermediate files in few posts up when you asked for them, the ll ones.

**Comment by JohanEngelen** [comment]
The ones generated by `-output-ll`? Those are the IR output by LDC. But that's before ThinLTO kicks in. "save-temps" gives you the IR output during+after ThinLTO.

**Comment by Fatmice** [comment]
> The ones generated by -output-ll? Those are the IR output by LDC. But that's before ThinLTO kicks in. "save-temps" gives you the IR output during+after ThinLTO.

Here we go.
[quyen@thinkpad lit-test]$ ../install/ldc/bin/ldc2 -v -L--plugin-opt=save-temps -O3 -flto=thin simple.d
binary    /home/quyen/D/install/ldc/bin/ldc2
version   1.4.0git-ea2b839 (DMD v2.074.1, LLVM 4.0.1)
config    /home/quyen/D/install/ldc/etc/ldc2.conf (x86_64-unknown-linux-gnu)
predefs   LDC all D_Version2 assert X86_64 D_InlineAsm_X86_64 D_HardFloat LittleEndian D_LP64 D_PIC linux Posix CRuntime_Glibc LDC_LLVM_400
parse     simple
importall simple
import    object	(/home/quyen/D/install/ldc/bin/../import/ldc/object.d)
semantic  simple
entry     main      	simple.d
semantic2 simple
semantic3 simple
code      simple
/usr/bin/gcc simple.o -o simple -Wl,-plugin,/home/quyen/D/install/ldc/lib64/LLVMgold-ldc.so -Wl,-plugin-opt=thinlto -Wl,-plugin-opt=O3 -Wl,-plugin-opt=-function-sections -Wl,-plugin-opt=-data-sections -L/home/quyen/D/install/ldc/bin/../lib64 -Xlinker --plugin-opt=save-temps -lphobos2-ldc -ldruntime-ldc -Wl,--gc-sections -lrt -ldl -lpthread -lm -m64 
[simple.zip](https://github.com/ldc-developers/ldc/files/1289080/simple.zip)

**Comment by Fatmice** [comment]
Without -O3
[quyen@thinkpad lit-test]$ ../install/ldc/bin/ldc2 -v -L--plugin-opt=save-temps -flto=thin simple.d
binary    /home/quyen/D/install/ldc/bin/ldc2
version   1.4.0git-ea2b839 (DMD v2.074.1, LLVM 4.0.1)
config    /home/quyen/D/install/ldc/etc/ldc2.conf (x86_64-unknown-linux-gnu)
predefs   LDC all D_Version2 assert X86_64 D_InlineAsm_X86_64 D_HardFloat LittleEndian D_LP64 D_PIC linux Posix CRuntime_Glibc LDC_LLVM_400
parse     simple
importall simple
import    object	(/home/quyen/D/install/ldc/bin/../import/ldc/object.d)
semantic  simple
entry     main      	simple.d
semantic2 simple
semantic3 simple
code      simple
/usr/bin/gcc simple.o -o simple -Wl,-plugin,/home/quyen/D/install/ldc/lib64/LLVMgold-ldc.so -Wl,-plugin-opt=thinlto -Wl,-plugin-opt=O0 -Wl,-plugin-opt=-function-sections -Wl,-plugin-opt=-data-sections -L/home/quyen/D/install/ldc/bin/../lib64 -Xlinker --plugin-opt=save-temps -lphobos2-ldc -ldruntime-ldc -Wl,--gc-sections -lrt -ldl -lpthread -lm -m64 

[simple-noOpt.zip](https://github.com/ldc-developers/ldc/files/1289082/simple-noOpt.zip)

**Comment by kinke** [comment]
After promote (step 1), this
```LL
@ldc.dso_slot = linkonce_odr hidden global i8* null
@ldc.dso_initialized = linkonce_odr hidden local_unnamed_addr global i8 0
define linkonce_odr hidden void @ldc.register_dso(i1, i8*) local_unnamed_addr #1
define linkonce_odr hidden void @ldc.dso_ctor.6simple()
define linkonce_odr hidden void @ldc.dso_dtor.6simple()
```
is `internalized` to
```LL
@ldc.dso_slot = internal global i8* null
@ldc.dso_initialized = internal local_unnamed_addr global i8 0
define internal void @ldc.register_dso(i1, i8*) local_unnamed_addr #1
define internal void @ldc.dso_ctor.6simple()
define internal void @ldc.dso_dtor.6simple()
```
I guess that accounts for the issue at hand.

**Comment by JohanEngelen** [comment]
Does removing `hidden` from `@ldc.dso_initialized` fix things?
(also for the dso_slot perhaps?)

**Comment by klickverbot** [comment]
`ldc.dso_initialized` shouldn't be `local_unnamed_addr`.

They very much have to be `hidden` (i.e. local to the ELF module in question) for things to work, but of course there might be an LTO bug with that.

**Comment by kinke** [comment]
@klickverbot: Can't we simply put `ldc.register_dso()` and the 2 globals into druntime, only emitting the global ctor/dtor into the object?

**Comment by klickverbot** [comment]
@kinke: `ldc.register_dso` is just an "outlined" version of the ctor/dtor code to reduce the code size. All those do is call `_d_dso_registry` from `druntime` with the appropriate parameters – the slot and other metadata local to the given ELF module (shared library/…). The whole point of `dso_slot` is for druntime to have a place in each shared library to store things.

**Comment by kinke** [comment]
Yeah I've read you nice comment (https://github.com/ldc-developers/ldc/blob/master/gen/modules.cpp#L391) in the meantime; maybe the weak reference approach would work now with LLVM 3.7+.

**Comment by kinke** [comment]
`local_unnamed_addr` is added by an optimization pass. So setting it explicitly to `none` doesn't change a thing, and setting it tentatively to `unnamed_addr` (not touched by optimizer) doesn't fix it.

**Comment by klickverbot** [comment]
Ah, seems like I misunderstood the semantics of `local_unnamed_addr`; looking at LangRef again, it should be fine.

If I understand correctly, the (thin) LTO main executable is linked against a conventional runtime library archive here, right? In this case, internalising `ldc.dso_initialized` (and hence dropping it from the symbol table) would lead to `_d_dso_registry` being called twice for the program image; once from the program's own ctor and once from the copy inside druntime.

> If this problem still occurs, then clearly it is not LLVM

Not necessarily – we are regularly finding bugs in LLVM that are not in the subset exercised by Clang. Of course, from a practical point of view we will still have to work around it on our side.

**Comment by kinke** [comment]
> If I understand correctly, the (thin) LTO main executable is linked against a conventional runtime library archive here, right? In this case, internalising ldc.dso_initialized (and hence dropping it from the symbol table) would lead to _d_dso_registry being called twice for the program image; once from the program's own ctor and once from the copy inside druntime.

Exactly. I've just tried with WeakODR linkage instead of LinkOnceODR, that keeps the `weak_odr hidden` instead of `internal` with -O3, but still doesn't fix the issue.

**Comment by kinke** [comment]
Note that the internalization happens with `-O3` in general, independent from LTO. So I'm not sure why this is working in the first place for non-LTO builds. @ximion filed 2 more issues about asserts in `_d_dso_register` (#2280, #2300). Maybe the linker and/or order of linker arguments plays a role.

A blocker for 1.4 final?

**Comment by JohanEngelen** [comment]
> Note that the internalization happens with -O3 in general, independent from LTO.

Really? (i cannot reproduce this on OSX)

**Comment by kinke** [comment]
Erm, I'm not sure what .ll files I checked lately, but internalization has always only happened as part of ThinLTO, sorry for the mixup.

**Comment by JohanEngelen** [comment]
OK, thanks clarification :)
I'm in a discussion with Teresa from LLVM to figure out what's broken. What linker is used? `gold`, `bfd`, something else?   (and could you test with different linkers? also with `lld` perhaps?)

**Comment by kinke** [comment]
Oh wow, I didn't think that would make a difference, but it does - I used the default Ubuntu 16.04 ld before (`GNU ld (GNU Binutils for Ubuntu) 2.26.1`). Using its gold (`GNU gold (GNU Binutils for Ubuntu 2.26.1) 1.11`) instead works: `bin/ldc2 -O3 -flto=thin -Xcc -fuse-ld=gold -run ../simple.d`.

**Comment by Fatmice** [comment]
There's no such -fuse-ld parameter for my ldc2...

**Comment by kinke** [comment]
Look carefully, I used the `-Xcc` prefix as it's a gcc argument (you can also use `-Xcc=-fuse-ld=gold`).

**Comment by JohanEngelen** [comment]
> Both gold and bfd generate binary that works on mine.

So then what linker is used per default on ubuntu? Hmm.

**Comment by JohanEngelen** [comment]
(so I'm thinking what is happening is that linker and LTO are not communicating correctly, such that LLVM-LTO thinks that the linked-to libraries do not contain `ldc.dso_slot` and `ldc.dso_initialized` symbols, which would then validate internalization of them)

**Comment by Fatmice** [comment]
> So then what linker is used per default on ubuntu? Hmm.

I'm on Fedora 26...
ld-linux-x86-64.so.2 ?
How do I tell?

**Comment by klickverbot** [comment]
`ld --version`?

**Comment by klickverbot** [comment]
(Or `gcc -L--version foo.o` to be sure.)

**Comment by klickverbot** [comment]
> (so I'm thinking what is happening is that linker and LTO are not communicating correctly, such that LLVM-LTO thinks that the linked-to libraries do not contain ldc.dso_slot and ldc.dso_initialized symbols, which would then validate internalization of them)

That's what I would think is happening here as well.

**Comment by Fatmice** [comment]
```[quyen@thinkpad lit-test]$ ../install/ldc/bin/ldc2 -flto=thin -L--version app.d 
collect2 version 7.1.1 20170622 (Red Hat 7.1.1-3)
/usr/bin/ld -plugin /usr/libexec/gcc/x86_64-redhat-linux/7/liblto_plugin.so -plugin-opt=/usr/libexec/gcc/x86_64-redhat-linux/7/lto-wrapper -plugin-opt=-fresolution=/tmp/ccqEgYLP.res -plugin-opt=-pass-through=-lgcc -plugin-opt=-pass-through=-lgcc_s -plugin-opt=-pass-through=-lc -plugin-opt=-pass-through=-lgcc -plugin-opt=-pass-through=-lgcc_s --build-id --no-add-needed --eh-frame-hdr --hash-style=gnu -m elf_x86_64 -dynamic-linker /lib64/ld-linux-x86-64.so.2 -o app /usr/lib/gcc/x86_64-redhat-linux/7/../../../../lib64/crt1.o /usr/lib/gcc/x86_64-redhat-linux/7/../../../../lib64/crti.o /usr/lib/gcc/x86_64-redhat-linux/7/crtbegin.o -L/home/quyen/D/install/ldc/bin/../lib64 -L/usr/lib/gcc/x86_64-redhat-linux/7 -L/usr/lib/gcc/x86_64-redhat-linux/7/../../../../lib64 -L/lib/../lib64 -L/usr/lib/../lib64 -L/home/intel/compilers_and_libraries_2017.0.098/linux/ipp/lib/intel64 -L/home/intel/compilers_and_libraries_2017.0.098/linux/compiler/lib/intel64 -L/home/intel/compilers_and_libraries_2017.0.098/linux/mkl/lib/intel64 -L/home/intel/compilers_and_libraries_2017.0.098/linux/tbb/lib/intel64/gcc4.7 -L/home/intel/compilers_and_libraries_2017.0.098/linux/daal/lib/intel64_lin -L/home/intel/compilers_and_libraries_2017.0.098/linux/ipp/lib/intel64 -L/home/intel/compilers_and_libraries_2017.0.098/linux/compiler/lib/intel64 -L/home/intel/compilers_and_libraries_2017.0.098/linux/mkl/lib/intel64 -L/home/intel/compilers_and_libraries_2017.0.098/linux/tbb/lib/intel64/gcc4.7 -L/home/intel/compilers_and_libraries_2017.0.098/linux/daal/lib/intel64_lin -L/home/intel/compilers_and_libraries_2017.0.098/linux/daal/../tbb/lib/intel64_lin/gcc4.4 -L/usr/lib/gcc/x86_64-redhat-linux/7/../../.. app.o -plugin /home/quyen/D/install/ldc/lib64/LLVMgold-ldc.so -plugin-opt=thinlto -plugin-opt=O0 -plugin-opt=-function-sections -plugin-opt=-data-sections --version -lphobos2-ldc -ldruntime-ldc --gc-sections -lrt -ldl -lpthread -lm -lgcc --as-needed -lgcc_s --no-as-needed -lc -lgcc --as-needed -lgcc_s --no-as-needed /usr/lib/gcc/x86_64-redhat-linux/7/crtend.o /usr/lib/gcc/x86_64-redhat-linux/7/../../../../lib64/crtn.o
GNU ld version 2.27-24.fc26
Copyright (C) 2016 Free Software Foundation, Inc.
This program is free software; you may redistribute it under the terms of
the GNU General Public License version 3 or (at your option) a later version.
This program has absolutely no warranty.
```

**Comment by Fatmice** [comment]
Correction, bfd does not work with optimization.
```
[quyen@thinkpad lit-test]$ ../install/ldc/bin/ldc2 -O3 -flto=thin -Xcc=-fuse-ld=bfd -run simple.d
Error: /tmp/simple-593002c-4f8cad failed with status: -2
Error: message: Illegal instruction (core dumped)
Error: program received signal 2 (Interrupt)
```

**Comment by Fatmice** [comment]
../install/ldc/bin/ldc2 -O3 -flto=thin -L--version -Xcc=-fuse-ld=gold -run simple.d
```
collect2 version 7.1.1 20170622 (Red Hat 7.1.1-3)
/usr/bin/ld.gold -plugin /usr/libexec/gcc/x86_64-redhat-linux/7/liblto_plugin.so -plugin-opt=/usr/libexec/gcc/x86_64-redhat-linux/7/lto-wrapper -plugin-opt=-fresolution=/tmp/ccySbWtG.res -plugin-opt=-pass-through=-lgcc -plugin-opt=-pass-through=-lgcc_s -plugin-opt=-pass-through=-lc -plugin-opt=-pass-through=-lgcc -plugin-opt=-pass-through=-lgcc_s --build-id --no-add-needed --eh-frame-hdr --hash-style=gnu -m elf_x86_64 -dynamic-linker /lib64/ld-linux-x86-64.so.2 -o /tmp/simple-00cb66b-1f57db /usr/lib/gcc/x86_64-redhat-linux/7/../../../../lib64/crt1.o /usr/lib/gcc/x86_64-redhat-linux/7/../../../../lib64/crti.o /usr/lib/gcc/x86_64-redhat-linux/7/crtbegin.o -L/home/quyen/D/install/ldc/bin/../lib64 -L/usr/lib/gcc/x86_64-redhat-linux/7 -L/usr/lib/gcc/x86_64-redhat-linux/7/../../../../lib64 -L/lib/../lib64 -L/usr/lib/../lib64 -L/home/intel/compilers_and_libraries_2017.0.098/linux/ipp/lib/intel64 -L/home/intel/compilers_and_libraries_2017.0.098/linux/compiler/lib/intel64 -L/home/intel/compilers_and_libraries_2017.0.098/linux/mkl/lib/intel64 -L/home/intel/compilers_and_libraries_2017.0.098/linux/tbb/lib/intel64/gcc4.7 -L/home/intel/compilers_and_libraries_2017.0.098/linux/daal/lib/intel64_lin -L/home/intel/compilers_and_libraries_2017.0.098/linux/ipp/lib/intel64 -L/home/intel/compilers_and_libraries_2017.0.098/linux/compiler/lib/intel64 -L/home/intel/compilers_and_libraries_2017.0.098/linux/mkl/lib/intel64 -L/home/intel/compilers_and_libraries_2017.0.098/linux/tbb/lib/intel64/gcc4.7 -L/home/intel/compilers_and_libraries_2017.0.098/linux/daal/lib/intel64_lin -L/home/intel/compilers_and_libraries_2017.0.098/linux/daal/../tbb/lib/intel64_lin/gcc4.4 -L/usr/lib/gcc/x86_64-redhat-linux/7/../../.. simple-00cb66b.o -plugin /home/quyen/D/install/ldc/lib64/LLVMgold-ldc.so -plugin-opt=thinlto -plugin-opt=O3 -plugin-opt=-function-sections -plugin-opt=-data-sections --version -lphobos2-ldc -ldruntime-ldc --gc-sections -lrt -ldl -lpthread -lm -lgcc --as-needed -lgcc_s --no-as-needed -lc -lgcc --as-needed -lgcc_s --no-as-needed /usr/lib/gcc/x86_64-redhat-linux/7/crtend.o /usr/lib/gcc/x86_64-redhat-linux/7/../../../../lib64/crtn.o
GNU gold (version 2.27-24.fc26) 1.12
Copyright (C) 2016 Free Software Foundation, Inc.
This program is free software; you may redistribute it under the terms of
the GNU General Public License version 3 or (at your option) a later version.
This program has absolutely no warranty.
Error: /tmp/simple-00cb66b-1f57db failed with status: -1
Error: message: Executable "/tmp/simple-00cb66b-1f57db" doesn't exist!
Error: program received signal 1 (Hangup)
```
will crash
but 
../install/ldc/bin/ldc2 -O3 -flto=thin -Xcc=-fuse-ld=gold -run simple.d
doesn't.

**Comment by JohanEngelen** [comment]
Can you try passing `-compute-dead=false` to the failing linker? (use something like `-L-Wl,plugin-opt=-compute-dead=false` )

**Comment by Fatmice** [comment]
```
[quyen@thinkpad lit-test]$ ../install/ldc/bin/ldc2 -O3 -flto=thin -L-Wl,plugin-opt="-compute-dead=false" -Xcc=-fuse-ld=gold -run simple.d
/usr/bin/ld.gold: error: cannot open plugin-opt=-compute-dead=false: No such file or directory
collect2: error: ld returned 1 exit status
Error: /usr/bin/gcc failed with status: 1
[quyen@thinkpad lit-test]$ ../install/ldc/bin/ldc2 -O3 -flto=thin -L-Wl,plugin-opt="-compute-dead=false" -Xcc=-fuse-ld=bfd -run simple.d
/usr/bin/ld.bfd: cannot find plugin-opt=-compute-dead=false: No such file or directory
collect2: error: ld returned 1 exit status
Error: /usr/bin/gcc failed with status: 1
[quyen@thinkpad lit-test]$ 
```

**Comment by JohanEngelen** [comment]
No quotes, `-L-Wl,-plugin-opt=-compute-dead=false` (extra minus)

**Comment by Fatmice** [comment]
```
[quyen@thinkpad lit-test]$ ../install/ldc/bin/ldc2 -v -O3 -flto=thin -L-Wl,-plugin-opt=-compute-dead=false -Xcc=-fuse-ld=bfd -run simple.d
binary    /home/quyen/D/install/ldc/bin/ldc2
version   1.4.0git-ea2b839 (DMD v2.074.1, LLVM 4.0.1)
config    /home/quyen/D/install/ldc/etc/ldc2.conf (x86_64-unknown-linux-gnu)
predefs   LDC all D_Version2 assert X86_64 D_InlineAsm_X86_64 D_HardFloat LittleEndian D_LP64 D_PIC linux Posix CRuntime_Glibc LDC_LLVM_400
parse     simple
importall simple
import    object	(/home/quyen/D/install/ldc/bin/../import/ldc/object.d)
semantic  simple
entry     main      	simple.d
semantic2 simple
semantic3 simple
code      simple
/usr/bin/gcc simple-d6a6ea5.o -o /tmp/simple-d6a6ea5-c65133 -Wl,-plugin,/home/quyen/D/install/ldc/lib64/LLVMgold-ldc.so -Wl,-plugin-opt=thinlto -Wl,-plugin-opt=O3 -Wl,-plugin-opt=-function-sections -Wl,-plugin-opt=-data-sections -L/home/quyen/D/install/ldc/bin/../lib64 -Wl,-plugin-opt=-compute-dead=false -fuse-ld=bfd -lphobos2-ldc -ldruntime-ldc -Wl,--gc-sections -lrt -ldl -lpthread -lm -m64 
/tmp/simple-d6a6ea5-c65133 
Error: /tmp/simple-d6a6ea5-c65133 failed with status: -2
Error: message: Illegal instruction (core dumped)
Error: program received signal 2 (Interrupt)
```

**Comment by JohanEngelen** [comment]
As far as I know, LLVM's (Thin)LTO is better tested/supported with gold than with bfd. In any case, we should file a bug report with either bfd or LLVM. It'd help to reduce the testcase, (C program is most helpful I think), but just our analysis so far should help too.

**Comment by JohanEngelen** [comment]
Too bad the `-compute-dead=false` didn't work.

**Comment by Fatmice** [comment]
> Too bad the -compute-dead=false didn't work.

No matter, at least -fuse-ld=gold is working with optimization.

So doesn't -fuse-ld=gold uses the system gold plugin under /usr/bin ?  How can you force it to use the one made by LLVM or copied by LDC (LLVMgold-ldc.so) ? Or I am confused?

**Comment by kinke** [comment]
Look at your own log output - LDC adds the path to its LTO plugin to the gcc command line.

**Comment by JohanEngelen** [comment]
I've added a note about ThinLTO+ld.bfd to the release notes' known issues. If you think that's giving it too much importance, please remove it.

**Comment by Fatmice** [comment]
Ah yes indeed.  I was reading the gcc manual for something like that, -iplugindir=
dir.  All of these emitted debug messages are merging into a haze.

**Comment by JohanEngelen** [comment]
I'm happy we found the cause (kind of) of this bug; thanks for the persistence guys.

**Comment by Fatmice** [comment]
So some update after I experimented more with build flags.
LLVM 6.0.0-svn was built with this monster
```
cmake -G Ninja -DCMAKE_INSTALL_PREFIX=/home/quyen/D/install/llvm -DLLVM_TARGETS_TO_BUILD="X86" -DLLVM_CCACHE_BUILD=ON -DLLVM_CCACHE_DIR=/home/quyen/D/llvm-build/pgo-ccache -DLLVM_BINUTILS_INCDIR=/usr/include -DLLVM_PROFDATA_FILE=/home/quyen/D/llvm-build/clang.profdata -DLLVM_PARALLEL_COMPILE_JOBS=4 -DLLVM_PARALLEL_LINK_JOBS=4 -DCMAKE_BUILD_TYPE=Release -DCMAKE_CXX_FLAGS_RELEASE="-O3" -DLLVM_LIBDIR_SUFFIX=64 -DLLVM_ENABLE_PROJECTS="libcxx;libcxxabi" -DCMAKE_C_COMPILER=/home/quyen/D/llvm-build/pgo/bin/clang -DCMAKE_CXX_COMPILER=/home/quyen/D/llvm-build/pgo/bin/clang++ -DCMAKE_RANLIB=/home/quyen/D/llvm-build/pgo/bin/llvm-ranlib -DCMAKE_AR=/home/quyen/D/llvm-build/pgo/bin/llvm-ar -DCMAKE_CXX_LINK_FLAGS="-Wl,-rpath,/home/quyen/D/install/llvm/lib64" -DCMAKE_SHARED_LINKER_FLAGS="-Wl,-rpath,/home/quyen/D/install/llvm/lib64" -DCMAKE_MODULE_LINKER_FLAGS="-Wl,-rpath,/home/quyen/D/install/llvm/lib64" -DLLVM_LINK_LLVM_DYLIB=ON -DLLVM_ENABLE_LLD=ON ../../llvm-svn/llvm
```
I'm now building LDC with LLVM 6.0.0-svn with
```
cmake -G Ninja -DCMAKE_INSTALL_PREFIX=../install/ldc -DINCLUDE_INSTALL_DIR=../install/ldc/import -DCMAKE_BUILD_TYPE=Release -DLIB_SUFFIX=64 -DCMAKE_CXX_FLAGS_RELEASE="-O3" -DCMAKE_C_COMPILER=/home/quyen/D/install/llvm/bin/clang -DCMAKE_CXX_COMPILER=/home/quyen/D/install/llvm/bin/clang++ -DCMAKE_RANLIB=/home/quyen/D/install/llvm/bin/llvm-ranlib -DCMAKE_AR=/home/quyen/D/install/llvm/bin/llvm-ar -DLDC_INSTALL_LTOPLUGIN=ON -DCMAKE_CXX_LINK_FLAGS="-Wl,-rpath,/home/quyen/D/install/llvm/lib64" -DCMAKE_EXE_LINKER_FLAGS="-Wl,-rpath,/home/quyen/D/install/llvm/lib64" -DLD_FLAGS="-Wl,-rpath,/home/quyen/D/install/ldc/lib64 -Wl,-rpath,'\$ORIGIN/../lib64'" -DLDC_WITH_LLD=OFF -DLDC_INSTALL_LLVM_RUNTIME_LIBS=ON -DLLVM_ROOT_DIR=../install/llvm -DD_COMPILER=../install/ldc-0.17.4-bootstrap/bin/ldmd2 ../ldc
```
Afterward, I can compile simple without the need to pass -Xcc -fuse-ld=gold
```
../install/ldc/bin/ldc2 -v -O3 -L-Wl,-rpath,/home/quyen/D/install/ldc/lib64 -flto=thin -run simple.d
binary    /home/quyen/D/install/ldc/bin/ldc2
version   1.4.0git-71a579d (DMD v2.074.1, LLVM 6.0.0svn)
config    /home/quyen/D/install/ldc/etc/ldc2.conf (x86_64-unknown-linux-gnu)
predefs   LDC all D_Version2 assert X86_64 D_InlineAsm_X86_64 D_HardFloat LittleEndian D_LP64 D_PIC linux Posix CRuntime_Glibc LDC_LLVM_600
parse     simple
importall simple
import    object	(/home/quyen/D/install/ldc/import/ldc/object.d)
semantic  simple
entry     main      	simple.d
semantic2 simple
semantic3 simple
code      simple
/usr/lib64/ccache/gcc simple-24644fa.o -o /tmp/simple-24644fa-373457 -Wl,-plugin,/home/quyen/D/install/ldc/lib64/LLVMgold-ldc.so -Wl,-plugin-opt=thinlto -Wl,-plugin-opt=O3 -Wl,-plugin-opt=-function-sections -Wl,-plugin-opt=-data-sections -Wl,-rpath,/home/quyen/D/install/ldc/lib64 -L/home/quyen/D/install/ldc/lib64 -lphobos2-ldc -ldruntime-ldc -Wl,--gc-sections -lrt -ldl -lpthread -lm -m64 
/tmp/simple-24644fa-373457
ldd simple
	linux-vdso.so.1 (0x00007fffa09ed000)
	libphobos2-ldc.so.74 => /home/quyen/D/install/ldc/lib64/libphobos2-ldc.so.74 (0x00007fb2d6806000)
	libdruntime-ldc.so.74 => /home/quyen/D/install/ldc/lib64/libdruntime-ldc.so.74 (0x00007fb2d64ea000)
	librt.so.1 => /lib64/librt.so.1 (0x00007fb2d62e2000)
	libdl.so.2 => /lib64/libdl.so.2 (0x00007fb2d60de000)
	libpthread.so.0 => /lib64/libpthread.so.0 (0x00007fb2d5ebf000)
	libm.so.6 => /lib64/libm.so.6 (0x00007fb2d5ba9000)
	libc.so.6 => /lib64/libc.so.6 (0x00007fb2d57d8000)
	/lib64/ld-linux-x86-64.so.2 (0x00007fb2d6edc000)
	libgcc_s.so.1 => /lib64/libgcc_s.so.1 (0x00007fb2d55c1000)
```
The libraries in ldc/lib64 are now mostly dynamic except for ldc-profile-rt, ldc-rt.asan/fuzzer which are archives.  Also a bonus is of course all of the binaries are much smaller now since druntime and phobos2 are now dynamically loaded to them.  Simple.d now compiles to binary of 8632 bytes whereas its C equivalent compiles to 7936 bytes.  This is huge improvement and departure from when it was 350864 bytes.

**Comment by kinke** [comment]
> Afterward, I can compile simple without the need to pass -Xcc -fuse-ld=gold

If I didn't miss anything, druntime/Phobos aren't compiled with LTO, so of course there should be no need to use gold instead of bfd.

> The libraries in ldc/lib64 are now mostly dynamic except for ldc-profile-rt, ldc-rt.asan/fuzzer which are archives. Also a bonus is of course all of the binaries are much smaller now since druntime and phobos2 are now dynamically loaded to them.

The shared druntime/Phobos libs should be created by default on Linux (as well as the static libs) - `libdruntime-ldc.a` and `libdruntime-ldc-shared.so`. So using `-defaultlib=druntime-ldc-shared,phobos2-ldc-shared -debuglib=druntime-ldc-debug-shared,phobos2-ldc-debug-shared` links your binary to the shared libs too, no need for building LDC in a special way.

**Comment by Fatmice** [comment]
llvm/trunk 313135 and ldc version 1.5.0git-b5c2656 doesn't seem to play nice...
[quyen@thinkpad lit-test]$ ../ldc-build/bin/ldc2 -v -O3 -L-Wl,-rpath,/home/quyen/D/install/ldc/lib64 -flto=thin -run simple.d
binary    /home/quyen/D/ldc-build/bin/ldc2
version   1.5.0git-b5c2656 (DMD v2.074.1, LLVM 6.0.0svn)
config    /home/quyen/D/ldc-build/bin/ldc2.conf (x86_64-unknown-linux-gnu)
predefs   LDC all D_Version2 assert X86_64 D_InlineAsm_X86_64 D_HardFloat LittleEndian D_LP64 D_PIC linux Posix CRuntime_Glibc LDC_LLVM_600
/home/quyen/D/install/llvm/lib64/libLLVM-6.0svn.so(+0x1bb3b09)[0x7f6c999e5b09]
/home/quyen/D/install/llvm/lib64/libLLVM-6.0svn.so(+0x1bb4006)[0x7f6c999e6006]
/lib64/libpthread.so.0(+0x123b0)[0x7f6c978193b0]
../ldc-build/bin/ldc2[0x3c6b54]
../ldc-build/bin/ldc2[0x45f2ce]
../ldc-build/bin/ldc2[0x54f9ce]
../ldc-build/bin/ldc2[0x48f17b]
../ldc-build/bin/ldc2[0x48d1a2]
../ldc-build/bin/ldc2[0x5f9641]
../ldc-build/bin/ldc2[0x51df5f]
../ldc-build/bin/ldc2[0x51df24]
../ldc-build/bin/ldc2[0x51de57]
/lib64/libc.so.6(__libc_start_main+0xea)[0x7f6c96ba150a]
../ldc-build/bin/ldc2[0x34202a]
Segmentation fault (core dumped)

Interestingly, ldc compiled against llvm/trunk 312880 works...

**Comment by kinke** [comment]
We only periodically check whether LLVM trunk is fine. It can be buggy at times too (not even compiling on Windows, I had that multiple times already).

On a side note, if your `-L-Wl,-rpath,/home/quyen/D/install/ldc/lib64` is required (not in `ldc2.conf`), that's a bug.

**Comment by Fatmice** [comment]
>On a side note, if your -L-Wl,-rpath,/home/quyen/D/install/ldc/lib64 is required (not in ldc2.conf), that's bug.

It's needed when I compiled LDC with BUILD_SHARED_LIBS=ON.  The libdruntime-ldc and libphobos2-ldc are shared but their path isn't in the system path (/usr/lib, /usr/lib64, etc) so I had to set rpath or runpath.

**Comment by kinke** [comment]
This shouldn't be the case, see https://github.com/ldc-developers/ldc/blob/master/runtime/CMakeLists.txt#L271. If you have some time, it'd be nice if you could have a quick look into it.

**Comment by Fatmice** [comment]
I can add that to my PR if you want and test it.

**Comment by kinke** [comment]
That'd be nice, but please open a new PR for it - I haven't merged your other PR yet because I'm not sure about the bash completions dir and the default dir you chose there (I don't think I've ever used these completions myself, so no idea).

**Comment by klickverbot** [comment]
```
commit 1111400a430ea72f7ca854cdbfe8769f0ab92e5b
Author: David Nadlinger <code@klickverbot.at>
Date:   Wed Jul 2 19:05:47 2014 +0200

    Set rpath for shared library test config.

    This way, the user doesn't have to set LD_LIBRARY_PATH to be
    able to run the test suite. We don't want to do this in the
    install config, though, as rpath has security implications.
```

**Comment by Fatmice** [comment]
@klickverbot So it was meant to be?

**Comment by kinke** [comment]
No idea about the security implications and what's common in the Linux world, but not having it set automatically for 'local' (~) LDC installations is quite a hassle IMO. Distro package maintainers can/should? always patch it out.

**Comment by klickverbot** [comment]
Having to manually take care of library paths (via LD_LIBRARY_PATH or setting rpaths during compilation) is absolutely normal in the Linux world.

The security problem comes from potentially ending up with references to libraries in paths controlled by other (especially less-privileged) users if directory structures change.

I believe Rust and others also converged on a similar solution – make sure the compiler and tooling itself can be run without manual intervention, but behave like a C compiler otherwise.

**Comment by kinke** [comment]
Alright, thx for the explanation and sorry for the noise then.

**Comment by klickverbot** [comment]
We could provide it as a CMake switch, though.

**Comment by Fatmice** [comment]
> We could provide it as a CMake switch, though.
Hm, yes that would be doable

**Comment by kinke** [comment]
We now default to the gold linker for ThinLTO on Linux and revamped the CLI interface (`-link-defaultlib-shared`, which is also the default for `-shared`, and which optionally sets the rpath according to the new config file setting etc.), so I think we can safely close this.
